### ISSUE TITLE
Got axes don't match array after writing the model

### ISSUE THREAD

**Issue by KelexBot** [autonomousvision/projected_gan#22]
Hi, 
I get the following issue while training the model on my dataset (just after the first trained model has been saved):

```
ValueError: Caught ValueError in DataLoader worker process 0.
Original Traceback (most recent call last):
  File "/usr/local/lib/python3.7/dist-packages/torch/utils/data/_utils/worker.py", line 287, in _worker_loop
    data = fetcher.fetch(index)
  File "/usr/local/lib/python3.7/dist-packages/torch/utils/data/_utils/fetch.py", line 49, in fetch
    data = [self.dataset[idx] for idx in possibly_batched_index]
  File "/usr/local/lib/python3.7/dist-packages/torch/utils/data/_utils/fetch.py", line 49, in <listcomp>
    data = [self.dataset[idx] for idx in possibly_batched_index]
  File "/content/projected_gan/training/dataset.py", line 102, in __getitem__
    image = self._load_raw_image(self._raw_idx[idx])
  File "/content/projected_gan/training/dataset.py", line 235, in _load_raw_image
    image = image.transpose(2, 0, 1) # HWC => CHW
ValueError: **axes don't match array**
```

I've first zipped my JPEG images with the dataset tools:

`python dataset_tool.py --source=xxx --dest=yyyy.zip --resolution=256x256
`

Note that I'm using MacOS so maybe it may impact the way the images are zipped...

I'm using the following command line:

`train(
    outdir='training-runs', 
    cfg='fastgan',
    data='xxxx, 
    gpus=1, 
    batch=64, 
    cond=False, 
    mirror=1, 
    batch_gpu=8, 
    cbase=32768, 
    cmax=512, 
    glr=None, 
    dlr=0.002, 
    desc='', 
    metrics=['fid50k_full'], 
    kimg=10000, 
    tick=4, 
    snap=1, 
    seed=0, 
    workers=0
)`

Thanks!

**Comment by xl-sr** [comment]
I think I've had this issue before. Can you try setting ```workers=2```?

**Comment by KelexBot** [comment]
Nope unfortunately:

With workers = 1: Bad magic number for file header
With workers = 2: Same issue

**Comment by xl-sr** [comment]
mhm strange. It is definitely a problem with the dataloading, but I don't think the zipping is the issue. You can also inspect the zip file and see if the images look as they should. I think I had this issue when setting up the colab, setting workers to 0 solved it.
Can you set workers to 0, put a breakpoint right before the buggy line, and print out the image shape?

**Comment by xl-sr** [comment]
I just pushed a commit that might solve your problem, can you try the dataset tool again? 

I noticed that in the standard StyleGAN dataset_tool the line ``` img = np.array(PIL.Image.open(fname))``` can lead to unexpected behaviour for custom datasets. Specifically, it sometimes drops the channel dimension for PNGs.

**Comment by KelexBot** [comment]
Nope unfortunately got the same issue.
Question => coud it be due to an alpha channel ?

**Comment by xl-sr** [comment]
it is likely due to the alpha channel, but that should be handled by imageio.

> Can you set workers to 0, put a breakpoint right before the buggy line, and print out the image shape?

What does this print?

**Comment by woctezuma** [comment]
Could you add a breakpoint or `print()` the shape of the image when it fails?

I suspect either an black-and-white image with one channel (which would have been squeezed) and ends up of shape HW instead of HWC.

**Comment by xl-sr** [comment]
yes, exactly. Are you using ```--transform=center-crop```? For this transform, I build in a black-and-white image check (since I am usually using this transform).

**Comment by vanga** [comment]
Hi,
I believe I am running into the same issue on Colab (though its not a colab specific issue).
The same dataset worked with SG2-ada-pytorch and SG3 repos.

- I have the latest cloned repo (as of yesterday)
- I tried multiple times and the same issue happens while trying to compute the metrics after the first checkpoint
- Training goes on if I disable metrics computation
- Tested by checking if all the images from the prepared dataset zip have RGB mode using PIL, all of them have RGB (I am not sure if I need to check this during training)
- Running calc_metrics later works though, issue seems to be when these metrics are being computed as part of the training. I used the pkl file from about 800 kimg to calculate the FID. I am not sure if the error is specific to first checkpoint.

Is there anything else that I can check to help debug this further?
